### PR TITLE
ignore CR in dumb terminal only when it is immediately preceeded by LF

### PR DESCRIPTION
**Pull request by hvesalai**
@gnodet The original code was, as I read the history, added to solve problems when piping text with CR-LFs to a process using jline. In such a a case, both CR and LF would trigger accept-line.

However completely ignoring CRs meant that line reader doesn't work for a dumb terminal whose carriage return really sends only CR. As I noted in #874.

### CODE REVIEW

**Comment by hvesalai** [comment]
As noted in https://github.com/jline/jline3/pull/874#issuecomment-1768249607, the current way to ignore CRs in dumb terminal renders the dumb terminal non-functional for real dumb terminal usage. This PR is an attempt to fix the shortcomings.

File: reader/src/main/java/org/jline/reader/impl/LineReaderImpl.java [hunk 1]
```diff
@@ -586,7 +590,11 @@ public String readLine(String prompt, String rightPrompt, MaskingCallback maskin
 
             smallTerminalOffset = 0;
 
-            state = State.NORMAL;
+            if (isDumb && "\n".equals(getLastBinding())) {
```
Comment (gnodet) [inline reader/src/main/java/org/jline/reader/impl/LineReaderImpl.java]: Could you add 2 constants for `CR = "\r"` and `LF = "\n"`, that would make the code more readable when talking about _ignoring CR_.

**Review by gnodet — APPROVED** [review]

File: reader/src/main/java/org/jline/reader/impl/LineReaderImpl.java (hunk 1 above)
Comment (hvesalai) [reply reader/src/main/java/org/jline/reader/impl/LineReaderImpl.java]: Added along with some more comments

File: reader/src/main/java/org/jline/reader/impl/LineReaderImpl.java [hunk 2]
```diff
@@ -586,7 +593,12 @@ public String readLine(String prompt, String rightPrompt, MaskingCallback maskin
 
             smallTerminalOffset = 0;
 
-            state = State.NORMAL;
+            if (isDumb && LF.equals(getLastBinding())) {
```
Comment (gnodet) [inline reader/src/main/java/org/jline/reader/impl/LineReaderImpl.java:596]: This sounds wrong.  This is the preparation of the `readLine` call, so we should not discard bindings here.  If there's a cr+lf in a previous call, it should have been consumed completely.

File: reader/src/main/java/org/jline/reader/impl/LineReaderImpl.java [hunk 3]
```diff
@@ -781,15 +788,15 @@ public void printAbove(String str) {
             if (reading) {
                 display.update(Collections.emptyList(), 0);
             }
-            if (str.endsWith("\n") || str.endsWith("\n\033[m") || str.endsWith("\n\033[0m")) {
+            if (str.endsWith(LF) || str.endsWith("\n\033[m") || str.endsWith("\n\033[0m")) {
```
Comment (gnodet) [inline reader/src/main/java/org/jline/reader/impl/LineReaderImpl.java]: I would leave that change out.  The `CR` and `LF` constants were meant to make it more obvious when dealing with those cr/lf issues.

File: reader/src/main/java/org/jline/reader/impl/LineReaderImpl.java (hunk 2 above)
Comment (gnodet) [reply reader/src/main/java/org/jline/reader/impl/LineReaderImpl.java:596]: See https://github.com/jline/jline3/pull/874/files#r1365449433

**Review by gnodet — CHANGES_REQUESTED** [review]
An ending `cr+lf` should be consumed completely at the end of the `readLine` call.

File: reader/src/main/java/org/jline/reader/impl/LineReaderImpl.java (hunk 2 above)
Comment (hvesalai) [reply reader/src/main/java/org/jline/reader/impl/LineReaderImpl.java:596]: See my comment on the other PR. I don't think that is possible.

Comment (hvesalai) [reply reader/src/main/java/org/jline/reader/impl/LineReaderImpl.java:596]: Note that we are not discarding anything there. Here we just check if the previous `ACCEPT_LINE` was triggered  by an `LF`. The `getLastBinding` returns `LF` if that is the case. It doesn't consume the binding.

**Comment by gnodet** [comment]
The LineReader is not supposed to cope with terminal specifics at the binding level.
I would agree with the removal of [those lines](https://github.com/jline/jline3/pull/875/files#diff-dddd64baa0e1929f2b9e8909cfc545a7b568aaa39d60b16c57a030fed004f0b5L636-L640):
```
                    // For dumb terminals, we need to make sure that CR are ignored
                    Attributes attr = new Attributes(originalAttributes);
                    attr.setInputFlag(Attributes.InputFlag.IGNCR, true);
                    terminal.setAttributes(attr);
 ```
but I think this responsibility should be moved to the `TerminalBuilder`.

I have hardly seen any terminal which sens `\r` only as a line break.  This is afaik only the case for Mac OS 9 or before.  I don't think they can even run JLine.  So we're left with either `\n` or `\r\n`.  The `IGNCR` flag should get rid of `\r`, which sounds ok to me.

What's your setup exactly ?  Does emacs actually send `\r` as line breaks when you hit `return` ?  If that's the case, I'd modify the `TerminalBuilder` to set up the terminal with `InputFlag.ICRNL` instead.  That would be broken, because the [terminal ignores the `ICRNL` flag when `IGNCR` is set](https://github.com/jline/jline3/blob/8685b2a6f2a7a00e7b026e0cbdd7a01ca73929d2/terminal/src/main/java/org/jline/terminal/impl/LineDisciplineTerminal.java#L257-L262).   This would mean that the terminal builder should either set up the dumb terminal with `InputFlag.ICRNL` flag if under emacs and `InputFlag.IGNCR` in other cases.

**Comment by hvesalai** [comment]
This problem (sending \r as a line break) has nothing to do with emacs. Emacs sends `\n`, so I' good if you merge the other PR.

The setup where I noticed the problem in this issue is (enter-key sends only `\r`):

- `Windows Subsystem for Linux 2`
- running `Ubuntu  22.04.3 LTS (GNU/Linux 5.15.90.1-microsoft-standard-WSL2 x86_64)`
- running `GNU bash, version 5.1.16(1)-release (x86_64-pc-linux-gnu)`
- used through `Windows Terminal 1.18.2822.0`

The way to reproduce this is:
- in bash prompt, type `env TERM=dumb; ./build demo` (or alternatively `export TERM=dumb` and then `./build demo`)

**Comment by gnodet** [comment]
> This problem (sending `\r` as a line break) has nothing to do with emacs. Emacs sends `\n`, so I' good if you merge the other PR.
> 
> The setup where I noticed the problem in this issue is (enter-key sends only `\r`):
> 
> * `Windows Subsystem for Linux 2`
> * running `Ubuntu  22.04.3 LTS (GNU/Linux 5.15.90.1-microsoft-standard-WSL2 x86_64)`
> * running `GNU bash, version 5.1.16(1)-release (x86_64-pc-linux-gnu)`
> * used through `Windows Terminal 1.18.2822.0`
> 
> The way to reproduce this is:
> 
> * in bash prompt, type `env TERM=dumb ./build demo` (or alternatively `export TERM=dumb` and then `./build demo`)
> 
> When doing so from the current `master`, the Enter-key doesn't result in any action because of the `IGNCR` for dumb terminals. Pressing `CTRL-J` or `CTRL-Enter` sends a `\n`, so that triggers the `ACCEPT_LINE`.

`env TERM=dumb ./build demo` is not a correct way to launch JLine.  This does not actually create a dumb terminal, but it will create the usual `org.jline.terminal.impl.PosixSysTerminal` but configured with a DUMB type.  The `PosixSysTerminal` assumes a non-dumb terminal to be configured.  I think in that case, the `TerminalBuilder` should create a dumb terminal whenever the `TERM` begins is `dumb` or starts with `dumb-`.

The terminal can be checked by entering `$.terminal` in the demo.

**Comment by gnodet** [comment]
See  https://github.com/jline/jline3/pull/877 as a possible fix

**Pull request closed without merging**